### PR TITLE
feat(engine-dom): introduce a renderer-factory that can produce a renderer

### PR DESCRIPTION
**Pull request by ravijayaramappa**
## Details
This PR introduces a renderer-factory api that can produce a renderer object. `@lwc/engine-dom` will have a base renderer that will be used for most cases. The renderer-factory adds the ability for containers to serialize the factory, evaluate it a sandbox and produce renderer that can then be injected using custom renderer hooks(https://github.com/salesforce/lwc/pull/2861).

**How this is suppose to work?**
When a template is compiled and custom renderer hook is added for eligible elements, the compiled code will have the following import.
```js
// compiled code
import { registerTemplate, renderer } from "lwc";
const stc0 = {...}; // ... for simplification
function tmpl($api, $cmp, $slotset, $ctx) {
  const { h: api_element } = $api;
  return [
    api_element("svg", stc0, [
      api_element("use", {
        attrs: {
          "xlink:href": $cmp.getXLink,
        },
        key: 1,
        svg: true,
        renderer: renderer,
      }),
    ]),
  ];
  /*LWC compiler vX.X.X*/
}
export default registerTemplate(tmpl);
tmpl.stylesheets = [];
```

The locker compiler plugin will sniff the `renderer` being imported and transform it further 
```js
import { renderer, rendererFactory } from 'lwc';

// PSUEDOCODE
const sandboxedRendererFactory = evaluateInSandbox(rendererFactory.toString());
const sandboxedRenderer = sandboxedRendererFactory(renderer);

const stc0 = {...};
function tmpl($api, $cmp, $slotset, $ctx) {
  const { h: api_element } = $api;
  return [
    api_element("svg", stc0, [
      api_element("use", {
        attrs: {
          "xlink:href": $cmp.getXLink,
        },
        key: 1,
        svg: true,
        renderer: sandboxedRenderer,
      }),
    ]),
  ];
  /*LWC compiler vX.X.X*/
}
export default registerTemplate(tmpl);
tmpl.stylesheets = [];
```

## Does this pull request introduce a breaking change?

* ✅ No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

* ✅ No, it does not introduce an observable change.

## GUS work item
W-11027860

### CODE REVIEW

File: packages/@lwc/engine-dom/src/renderer-factory.ts [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+        return obj === null;
+    }
+
+    function isUndefined(obj: any): obj is undefined {
+        return obj === undefined;
+    }
+
+    // Note: These keys have to be kept in sync with @lwc/shared
+    const KEY__IS_NATIVE_SHADOW_ROOT_DEFINED = '$isNativeShadowRootDefined$';
+    const KEY__SHADOW_TOKEN = '$shadowToken$';
```
Comment (nolanlawson) [inline packages/@lwc/engine-dom/src/renderer-factory.ts:50]: Why duplicate these properties from `@lwc/shared`?

Comment (ravijayaramappa) [reply packages/@lwc/engine-dom/src/renderer-factory.ts:48]: The renderer-factory needs to be a pure function without any outside dependencies(imports). The goal is that it will be toString-ed and eval-ed inside a sandbox. Invocations the sandboxed renderer-factory will produce a renderer.

File: packages/@lwc/engine-core/src/framework/renderer.ts [hunk 2]
```diff
@@ -24,7 +24,7 @@ export interface RendererAPI {
     nextSibling: (node: N) => N | null;
     attachShadow: (element: E, options: ShadowRootInit) => N;
     getProperty: (node: N, key: string) => any;
-    setProperty: (node: N, key: string, value: any) => void;
+    setProperty: (node: N, key: string, value: any, attributeName: string) => void;
```
Comment (caridy) [inline packages/@lwc/engine-core/src/framework/renderer.ts:27]: I don't understand why the new argument is now required?

File: packages/@lwc/engine-dom/src/renderer.ts [hunk 3]
```diff
@@ -5,314 +5,8 @@
...
-    getFirstElementChild,
-    getLastChild,
-    getLastElementChild,
-    isConnected,
-    insertStylesheet,
-    assertInstanceOfHTMLElement,
-    defineCustomElement,
-    getCustomElement,
-};
+export const renderer = rendererFactory(globalThis, insertStylesheet);
```
Comment (caridy) [inline packages/@lwc/engine-dom/src/renderer.ts:12]: how would locker be able to import `insertStylesheet`?

File: packages/@lwc/engine-core/src/framework/renderer.ts (hunk 2 above)
Comment (ravijayaramappa) [reply packages/@lwc/engine-core/src/framework/renderer.ts:27]: There is validation in `setProperty` implementation that verifies if the property is supported by the element. The error message needs the `htmlPropertyToAttribute`(implemented in `@lwc/shared`) conversion. 
Instead of `@lwc/engine-dom` renderer relying on this function from `@lwc/shared`, the engine-core will get value and pass it to `setProperty`.

File: packages/@lwc/engine-dom/src/renderer.ts (hunk 3 above)
Comment (ravijayaramappa) [reply packages/@lwc/engine-dom/src/renderer.ts:12]: We discussed doing it in a follow up PR.

Comment (ravijayaramappa) [reply packages/@lwc/engine-dom/src/renderer.ts:12]: We discussed doing it in a follow up PR.

Comment (ravijayaramappa) [reply packages/@lwc/engine-dom/src/renderer.ts]: Decided to just do it in this PR.

Comment (ravijayaramappa) [reply packages/@lwc/engine-dom/src/renderer.ts]: @caridy We had talked about doing something like 
```
rendererFactory.static.createStyleInsertion(src) {
    // ...
}
```
I figured all of that is not required. We can just do 
```
import { renderer, rendererFactory } from 'lwc';
const sandboxedRendererFactory = evaluateInSandbox(rendererFactory.toString());
const sandboxedRenderer = sandboxedRendererFactory(globalThis);
sandboxedRenderer.__proto__ = renderer;
```
The sandboxedRenderer will inherit all missing properties from `renderer`.

File: packages/@lwc/engine-core/src/framework/renderer.ts (hunk 2 above)
Comment (caridy) [reply packages/@lwc/engine-core/src/framework/renderer.ts:27]: let's chat about it... there has to be another way.

File: packages/@lwc/engine-core/src/framework/modules/props.ts [hunk 4]
```diff
@@ -43,6 +43,20 @@ export function patchProps(
             isFirstPatch ||
             cur !== (isLiveBindingProp(sel, key) ? getProperty(elm!, key) : oldProps[key])
         ) {
+            // Additional verification if properties are supported by the element
```
Comment (ravijayaramappa) [inline packages/@lwc/engine-core/src/framework/modules/props.ts:46]: Moved this validation from [@lwc/engine-dom/src/renderer.ts](https://github.com/salesforce/lwc/pull/2997/files#diff-d700e3ea8d9babd8109d1b3e94e40cf4f45231cfd3ab2a47b760ca7dfff3d3bc)

File: packages/@lwc/engine-core/src/framework/hydration.ts [hunk 5]
```diff
@@ -524,15 +524,15 @@ function areCompatibleNodes(client: Node, ssr: Node, vnode: VNode, renderer: Ren
...
+        return getProperty(client, NODE_VALUE_PROP) === getProperty(ssr, NODE_VALUE_PROP);
     }
 
     if (getProperty(client, 'nodeType') === EnvNodeTypes.COMMENT) {
         if (!hasCorrectNodeType(vnode, ssr, EnvNodeTypes.COMMENT, renderer)) {
             return false;
         }
 
-        return getProperty(client, 'nodeValue') === getProperty(ssr, 'nodeValue');
+        return getProperty(client, NODE_VALUE_PROP) === getProperty(ssr, NODE_VALUE_PROP);
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/hydration.ts:535]: The changes in this file seem unrelated to this PR?

File: packages/@lwc/engine-core/src/framework/modules/props.ts [hunk 6]
```diff
@@ -43,6 +43,20 @@ export function patchProps(
...
+                if (!(key in elm!)) {
+                    assert.fail(
+                        `Unknown public property "${key}" of element <${
+                            elm!.tagName
+                        }>. This is likely a typo on the corresponding attribute "${htmlPropertyToAttribute(
+                            key
+                        )}".`
+                    );
+                }
+            }
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/modules/props.ts:59]: Why was this moved and put behind a `process.env.IS_BROWSER` check?

File: packages/@lwc/engine-dom/src/renderer-factory.ts [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+ * runtime environment. In the case of @lwc/enigne-dom, it is meant to be used in a DOM environment.
+ * Example usage:
+ * import { renderer, rendererFactory } from 'lwc';
+ * const customRenderer = rendererFactory(renderer);
+ *
+ * @param baseRenderer Either null or the base renderer imported from 'lwc'.
+ */
+export function rendererFactory(
+    baseRenderer: RendererAPI | null
+): Omit<RendererAPI, 'insertStylesheet' | 'isNativeShadowDefined' | 'isSyntheticShadowDefined'> {
```
Comment (nolanlawson) [inline packages/@lwc/engine-dom/src/renderer-factory.ts:26]: It seems like `insertStylesheet` is potentially something Locker may want to override? It has some security considerations (albeit minor ones).

Comment (nolanlawson) [reply packages/@lwc/engine-dom/src/renderer-factory.ts:26]: Nevermind, I see the comment below about the global cache.

File: packages/@lwc/engine-core/src/framework/hydration.ts (hunk 5 above)
Comment (ravijayaramappa) [reply packages/@lwc/engine-core/src/framework/hydration.ts:535]: Yes, it is unrelated. I had made a change in this file but reverted that. Using the constant was part of that change, I left this part in because it was useful.

File: packages/@lwc/engine-core/src/framework/modules/props.ts (hunk 6 above)
Comment (ravijayaramappa) [reply packages/@lwc/engine-core/src/framework/modules/props.ts:59]: This validation relies on `key in elm` check to work, where key is handled by descriptors for the element's global html properties and public(`@api`) properties. These properties are only available on the element in a browser environment and not in SSR.

**Comment by ravijayaramappa** [comment]
@nolanlawson @caridy This is ready for another pass.

File: packages/@lwc/engine-dom/src/renderer-factory.ts [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+            let content: Node = doc.body;
+            if (!isUndefined(wrapperTags)) {
+                for (let i = 0; i < wrapperTags.length; i++) {
+                    content = content.firstChild!;
+                }
+            }
+
+            return content.firstChild;
+        };
+    }
```
Comment (nolanlawson) [inline packages/@lwc/engine-dom/src/renderer-factory.ts:184]: I see it as a downside of this PR that we are now required to move all these functions into one big file. Conceptually, I think it was nice for `createFragment` to live in its own file.

Comment (ravijayaramappa) [reply packages/@lwc/engine-dom/src/renderer-factory.ts:184]: Talked to Nolan about this offline, we have some ideas of creating a separate `@lwc/renderer-dom` package where the renderer code can be modularized and using rollout to create the renderer-factory. Will tackle that as a separate follow up PR.

Comment (caridy) [reply packages/@lwc/engine-dom/src/renderer-factory.ts:184]: That was the original idea... but with the new twist, initially we were saying that locker will depend on that package directly to roll it into lockers source... but with the factory, we can have the best of both worlds.

Comment (nolanlawson) [reply packages/@lwc/engine-dom/src/renderer-factory.ts:184]: I thought about this some more... I'm not sure we even need separate packages. My main concern is that Locker implementation details don't bleed too much into LWC's source code. If we split the renderer APIs into their own file that exports top-level functions, and if we have a build step in `@lwc/engine-dom`, then I think we can do it all in one package without extra packages needed.

A good example of why we need this is pivots (#2724) – there is no way we want to put all that code inside of a single function with no external dependencies. 🙂

File: packages/@lwc/engine-dom/src/renderer-factory.ts [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+        getFirstElementChild,
+        getLastChild,
+        getLastElementChild,
+        isConnected,
+        assertInstanceOfHTMLElement,
+        defineCustomElement,
+        getCustomElement,
+    };
+    // Meant to inherit any properties passed via the base renderer as the argument to the factory.
+    Object.setPrototypeOf(renderer, baseRenderer);
```
Comment (abdulsattar) [inline packages/@lwc/engine-dom/src/renderer-factory.ts:384]: Can you explain why we are doing the inheritance here? 

From what I understand, locker should be able to override these methods in the renderer with lockerized versions of those (which are in `baseRenderer`). However based on this line, The methods in `renderer` will get invoked rather than the ones in `baseRenderer`.

**Review by abdulsattar — APPROVED** [review]

**Review by caridy — APPROVED** [review]
LGTM
